### ISSUE TITLE
updating server configuration 

### ISSUE THREAD

**Issue by Vathasav** [EUDAT-B2SAFE/B2SAFE-core#94]
Configuration:
irods_version - 4.2.1
CentOS - 7

While updating the server configuration, the install.sh script seems to corrupt the json. The updated rule base set is 
```json
"re_rulebase_set":["core","regexes_for_supported_peps":["ac[^]*","msi[^]*","[^]*pep_[^]*_(pre|post)"]},"shared_memory_instance":"irods_rule_language_rule_engine"},{"instance_name":"irods_rule_engine_plugin-cpp_default_policy-instance","plugin_name":"irods_rule_engine_plugin-cpp_default_policy","plugin_specific_configuration":{}}],{"filename":"eudat0"},{"filename":"eudat1"},{"filename":"eudat2"},{"filename":"eudat3"},{"filename":"eudat4"}], 
```

It has to be something like

```json
"re_rulebase_set":["core",{"regexes_for_supported_peps":["ac[^]*","msi[^]*","[^]*pep_[^]*_(pre|post)"]},{"shared_memory_instance":"irods_rule_language_rule_engine"},{"instance_name":"irods_rule_engine_plugin-cpp_default_policy-instance","plugin_name":"irods_rule_engine_plugin-cpp_default_policy","plugin_specific_configuration":{}},{"filename":"eudat0"},{"filename":"eudat1"},{"filename":"eudat2"},{"filename":"eudat3"},{"filename":"eudat4"}],
```
Irods server and utility scripts (e.g. irodsctl) fail because of this error.

**Comment by cookie33** [comment]
This probably has to do with the following:
```
       # put json file in a single string so it easier to process.
       EUDAT_SERVER_CONFIG_JSON_STRING=`awk 1 ORS=' ' ${EUDAT_SERVER_CONFIG} | sed 's/ //g'`
       JSON_RULESET=`echo $EUDAT_SERVER_CONFIG_JSON_STRING | sed 's/.*re_rulebase_set":/re_rulebase_set":/' | sed 's/}].*$/}]/'`
```
The whole json file is put in a single string at step 1.
The `re_rulebase_set` is extracted from the string.

This has been tested and worked on 4.x. But at 4.2.1 the string contains: 
```
{"regexes_for_supported_peps":["ac[^]*","msi[^]*","[^]*pep_[^]*_(pre|post)"]}
```
Because of the regex in here it might fail.

There is a before and after file of the `/etc/irods/server_config.json`. Can you post the differences? The before has a date timestamp.

**Comment by Vathasav** [comment]
This is the old configuration

```json

                   "re_data_variable_mapping_set": [
                        "core"
                    ],
                    "re_function_name_mapping_set": [
                        "core"
                    ],
                    "re_rulebase_set": [
                        "core"
                    ],
                    "regexes_for_supported_peps": [
                        "ac[^ ]*",
                        "msi[^ ]*",
                        "[^ ]*pep_[^ ]*_(pre|post)"
                    ]
                    
```

New configuration seems to be adding a new key instead of modifying the existing line:

```json

                    "re_data_variable_mapping_set": [
                        "core"
                    ],
                    "re_function_name_mapping_set": [
                        "core"
                    ],
                    "re_rulebase_set":["core",{"regexes_for_supported_peps":["ac[^]*","msi[^]*","[^]*pep_[^]*_(pre|post)"]},{"shared_memory_instance":"irods_rule_language_rule_engine"},{"instance_name":"irods_rule_engine_plugin-cpp_default_policy-instance","plugin_name":"irods_rule_engine_plugin-cpp_default_policy","plugin_specific_configuration":{}},{"filename":"eudat0"},{"filename":"eudat1"},{"filename":"eudat2"},{"filename":"eudat3"},{"filename":"eudat4"}],
                    "regexes_for_supported_peps": [
                        "ac[^ ]*",
                        "msi[^ ]*",
                        "[^ ]*pep_[^ ]*_(pre|post)"
                    ]
                
```

**Comment by cookie33** [comment]
This is how it looks in iRODS v4.1:
```
    "re_rulebase_set":[{"filename":"core"},{"filename":"eudat0"},{"filename":"eudat1"},{"filename":"eudat2"},{"filename":"eudat3"},{"filename":"eudat4"}], 
```
So the structure has changed. We need to adapt the replacement better for iRODS 4.2.1.

**Comment by cookie33** [comment]
And this is the problem:
```
                    "re_rulebase_set": [
                        "core"
                    ],
```
The search/replace looks for `[{` and `}]` which is not there in 4.2.1 it seems.

**Comment by cookie33** [comment]
And the result should have been:
```
                   "re_data_variable_mapping_set": [
                        "core"
                    ],
                    "re_function_name_mapping_set": [
                        "core"
                    ],
                    "re_rulebase_set": [
                        "core"
                    ],
                    "regexes_for_supported_peps": [
                        "ac[^ ]*",
                        "msi[^ ]*",
                        "[^ ]*pep_[^ ]*_(pre|post)"
                    ]
```

But the documentation of iRODS 4.2.1  https://docs.irods.org/4.2.1/system_overview/configuration/  states:

re_rulebase_set (required) - An array of file names comprising the list of rule files used by the rule engine, for example: { "filename": "core" } which references 'core.re'. This array is ordered as the order of the rule files affects which (multiply) matching rule would fire first.

If this was the case it would have worked. So Is there a change not reflected in the documentation?

**Comment by cookie33** [comment]
The user group meeting slide http://slides.com/irods/ugm2017-rule-engine-plugins#/4 give the new structure. So this seems to be valid..

Made an issue at the iRODS consortium  https://github.com/irods/irods/issues/3597

**Comment by cookie33** [comment]
Implementation proposal:
modify `update_server_config_json()`
put whole json file in one string as before.
* if `"filename"` exists in string of server_config.json we have old format and behave as before.
* if NO `"filename"` exists in string of server_config.json implement this new behaviour. Than we have 4.2.1 or higher of iRODS.

**Comment by cookie33** [comment]
Hi,

There is a fix in pull request #95.

I tested and it works on my machine for both old and new formats

Are you able to test it and confirm that it works? The plan than is (as the user who runs iRODS):
```
cd /etc/irods
cp server_config.json.20170707  server_config.json
cd /opt/eudat/b2safe/packaging
source /etc/irods/service_account.config
./install.sh
```

**Comment by cookie33** [comment]
Fixed in 4.0-1.
